new branch populate

Taktar opened the pull request:
Fixes #157 (Populate "Impact Created" with real data)

**What was changed?**

1. Shifts.js:
	◦	Added utility functions (calculateTotalHours, calculateUniqueShelters) for calculating data from past shifts.
	◦	Updated PastShifts to fetch past shifts, compute the required data, and pass it back to the parent using onImpactDataUpdate.

2. VolunteerDashboard.js:
	◦	Added useState to store totalHours and sheltersServed dynamically.
	◦	Passed onImpactDataUpdate to PastShifts to dynamically update the "Impact Created" section with calculated data.

**Why was it changed?**

This approach ensures that the "Impact Created" section dynamically displays real data fetched from the API. The "Impact Created" section on the Volunteer Dashboard previously displayed hardcoded values, which did not reflect the actual contributions of the volunteers. This issue was addressed to:
- Provide accurate, dynamic metrics (Total Hours Served and Shelters Served) based on past shifts.
- Enhance the user experience by showing real-time data that updates automatically when changes occur in the backend.

**How was it changed?**

Shifts.js :
Added Utility Functions:
- calculateTotalHours: Loops through past shifts, calculates the duration of each shift, and sums them up to provide the total hours served.
- calculateUniqueShelters: Uses a Set to count the number of unique shelters a volunteer has worked at.

Modified PastShifts Component:
- Fetches past shifts from the /shifts?filter_end_before=<time_now> API endpoint.
- Uses the above utility functions to calculate totalHours and sheltersServed.
-  Sends the calculated data back to the parent component using the onImpactDataUpdate callback.

VolunteerDashboard.js
Added useState:
- Used useState to manage totalHours and sheltersServed in the "Impact Created" section.
- Connected PastShifts to Update Dashboard:
Passed the onImpactDataUpdate callback to the PastShifts component.
- Updated the "Impact Created" section dynamically with the received data.
Replaced hardcoded values in the "Impact Created" section with dynamic values from impactData.

**Screenshots that show the changes (if applicable):**

Taktar commented:
<img width="1434" alt="Screenshot 2025-01-30 at 1 00 24 PM" src="https://github.com/user-attachments/assets/b147152a-c43e-4054-9da8-b6ecc331a711" />

Taktar commented:
<img width="1434" alt="Screenshot 2025-01-30 at 1 00 24 PM" src="https://github.com/user-attachments/assets/4b8abc2f-c9e5-4155-be2c-df9353e98ade" />

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -49,13 +64,28 @@ export function UpcomingShifts() {
   return (
     <div>
       <h1 className="text-center">Previous Shifts</h1>
-      {Shifts(shelters_endpoint)}
```
I think this changed causes "Past Shifts" not to show up in the volunteer dashboard.

kate-holdener commented on client_app/src/components/volunteer/VolunteerDashboard.js:
```diff
@@ -23,20 +30,21 @@ function VolunteerDashboard() {
       <div className="cta-button">
         <button onClick={signupClick}>Sign up for shifts</button>
       </div>
+      <PastShifts onImpactDataUpdate={(data) => setImpactData(data)} />
```
This change places "Past Shifts" into the dashboard. I don't think we want to do that.
<img width="1533" alt="Screenshot 2025-01-30 at 3 20 42 PM" src="https://github.com/user-attachments/assets/e901dfe6-7e10-4d39-9588-fbfdb754c91b" />

kate-holdener commented on client_app/src/components/volunteer/VolunteerDashboard.js:
```diff
@@ -23,20 +30,21 @@ function VolunteerDashboard() {
         <h3>Impact Created</h3>
         <div className="card">
           <h4>Total hours served</h4>
-          <p>106</p>
+          <p>{impactData.totalHours}</p>
         </div>
         <div className="card">
           <h4>Lives Touched</h4>
-          <p>17</p>
+          <p>Too many to count</p>
         </div>
         <div className="card">
           <h4>Shelters served</h4>
-          <p>4</p>
+          <p>{impactData.sheltersServed}</p>
```
Maybe all of this can be moved to Impact.js, where you call the backend API to retrieve the past shifts and calculate the hours and shelters.

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -3,13 +3,27 @@ import { SERVER } from "../../config";
 import ShiftList from "./ShiftList";
 import getAuthHeader from "../../authentication/getAuthHeader";
 
+// Utility functions for calculations
+const calculateTotalHours = (shifts) => {
+  return shifts.reduce((total, shift) => {
+    const start = new Date(shift.start_time);
+    const end = new Date(shift.end_time);
+    return total + Math.round((end - start) / (1000 * 60 * 60)); // Convert milliseconds to hours
+  }, 0);
+};
+
+const calculateUniqueShelters = (shifts) => {
+  const uniqueShelters = new Set(shifts.map((shift) => shift.shelter_id));
+  return uniqueShelters.size;
+};
+
```
I think that instead of modifying the Shifts module, you should create a new file Impact.js, for example. Move the calculation logic there.

strejoarciles replied:
I noticed that you ended up creating a new Impact.js file, I think this makes things neater.

kate-holdener commented:
@Taktar I added some comments explaining some issues with your changes. In summary, instead of modifying Shifts.js, I think you should create a new module Implact.js, where you call the backend API for the past shifts data and then calculate the hours and shelters served based on the data. You can move the rendering of that data there as well (from Dashboard.js) and then call on the Impact module from Dashboard.

Taktar commented:
[like]  Taslima Aktar reacted to your message:
________________________________
From: Kate Holdener ***@***.***>
Sent: Thursday, January 30, 2025 9:22:48 PM
To: oss-slu/shelter_volunteers ***@***.***>
Cc: Taslima Aktar ***@***.***>; Assign ***@***.***>
Subject: [External] Re: [oss-slu/shelter_volunteers] new branch populate (PR #161)

@ kate-holdener commented on this pull request. In client_app/src/components/volunteer/Shifts. js: > +// Utility functions for calculations +const calculateTotalHours = (shifts) => { + return shifts. reduce((total, shift) => { + const


@kate-holdener commented on this pull request.

________________________________

In client_app/src/components/volunteer/Shifts.js<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/161*discussion_r1936314586__;Iw!!K543PA!IzUvnQKwjilp7OKCVqvh2djgjfhElA_gQeBEZXoY53G0gWT9RTPrzpO94XuB3q2xdI6CEMbQ88KDlivNDMqRW6NSwhs$>:

> +// Utility functions for calculations
+const calculateTotalHours = (shifts) => {
+  return shifts.reduce((total, shift) => {
+    const start = new Date(shift.start_time);
+    const end = new Date(shift.end_time);
+    return total + Math.round((end - start) / (1000 * 60 * 60)); // Convert milliseconds to hours
+  }, 0);
+};
+
+const calculateUniqueShelters = (shifts) => {
+  const uniqueShelters = new Set(shifts.map((shift) => shift.shelter_id));
+  return uniqueShelters.size;
+};
+


I think that instead of modifying the Shifts module, you should create a new file Impact.js, for example. Move the calculation logic there.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/oss-slu/shelter_volunteers/pull/161*pullrequestreview-2585087591__;Iw!!K543PA!IzUvnQKwjilp7OKCVqvh2djgjfhElA_gQeBEZXoY53G0gWT9RTPrzpO94XuB3q2xdI6CEMbQ88KDlivNDMqRcn1dGjs$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/BFOPC4GTF5B5E3PYUGAPAG32NKJ2RAVCNFSM6AAAAABWDY6ID6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOBVGA4DONJZGE__;!!K543PA!IzUvnQKwjilp7OKCVqvh2djgjfhElA_gQeBEZXoY53G0gWT9RTPrzpO94XuB3q2xdI6CEMbQ88KDlivNDMqR47McrZI$>.
You are receiving this because you were assigned.Message ID: ***@***.***>

kate-holdener reviewed (changes requested):
See my comments for requested changes.

kate-holdener commented on client_app/src/components/volunteer/Impact.js:
This module is great: it wraps together the calculations we need and the rendering of the calculations in one reusable component.

kate-holdener commented on client_app/src/components/volunteer/Impact.js:
```diff
@@ -0,0 +1,55 @@
+      .then((response) => response.json())
+      .then((shifts) => {
+        if (shifts && shifts.length > 0) {
+          const totalHours = shifts.reduce((acc, shift) => {
+            const start = new Date(shift.start_time);
+            const end = new Date(shift.end_time);
+            return acc + Math.round((end - start) / (1000 * 60 * 60)); // Convert to hours
+          }, 0);
+
+          const uniqueShelters = new Set(shifts.map((shift) => shift.shelter));
```
You could improve this by moving this code into two separate functions: one for calculating hours and one for calculating shelters served. Then, you can call these functions from here.

strejoarciles replied:
The differentiation of the two calculations could lead to easier to call these separate functions.

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
Since you moved the hours and number of shelters calculation into a separate module (Impact), you should undo the changes on the Shifts module. The Shifts module is responsible for presenting upcoming and past shifts. It's unrelated to calculating the impact (even though Impact and Shifts use the same data, their responsibilities are unrelated). Here's a reference on how to restore one file to a previous version: https://graphite.dev/guides/revert-file-git

In this case, the commit has we want to restore this file to is: bc9f8900a13ed (this is the commit right before your changes). So the command you would use to restore this file is:
```
git restore --source bc9f8900a13ed client_app/src/components/volunteer/Shifts.js
```

kate-holdener reviewed (changes requested):
@Taktar Your changes are great and it looks like they are working. My json data that is loaded into the database contains 3 shifts. Each shift is one hour long and each shift is at a different shelter. Here's the result I get:
<img width="1538" alt="Screenshot 2025-01-31 at 8 46 30 AM" src="https://github.com/user-attachments/assets/8ddf83f7-6706-47da-952e-a8e441e664ad" />

When I go to "Your Previous Shifts", they don't show up because of the changes to the Shifts component. You can simply undo those changes (I included a comment on how to do that). I also suggested an improvement to the Impact module (it works as is, my suggestion is an improvement on organizing the code).

Great job!

Taktar commented:
1. Moving Calculations to Impact.js
Why: The original comments suggested that calculations for total hours served and shelters served should not be in Shifts.js because it is meant for managing shifts, not calculating impact. By creating Impact.js, you separated concerns, making the code more modular and easier to maintain.
How: Moved the calculation logic for total hours and unique shelters to Impact.js.
Ensured Impact.js fetches past shift data and performs the calculations internally.
Updated VolunteerDashboard.js to import and use the Impact component.

2. Fixing Shifts.js
Why: Changes to Shifts.js caused issues where shifts were not displaying properly (e.g., past shifts not showing). The comments pointed out that Shifts.js should remain focused on fetching and displaying shifts, not calculating impact.
How:
Reverted unnecessary changes to Shifts.js.
Removed calculation logic and restored it to fetch and display shifts correctly.
Updated its usage in VolunteerDashboard.js to ensure proper rendering.

3. Updating VolunteerDashboard.js
Why: The dashboard needed to properly integrate Impact.js and ensure the sections (Upcoming Shifts, Shelters, Impact) were modular and functional. Comments also highlighted JSX formatting issues.
How:
Added Impact as a standalone component to display calculated data.
Fixed JSX warnings by placing JSX elements on the same line where needed (e.g., <h3> with its component).
Ensured navigation (signupClick) and all sections work as expected.

4. General Improvements
Why: To address modularity, maintainability, and code clarity concerns from comments.
How:
Split responsibilities: Impact.js handles calculations, Shifts.js handles shifts, and VolunteerDashboard.js organizes the dashboard layout.
Fixed ESLint warnings (e.g., JSX formatting, proper imports).


***** I am providing screenshots of what I got before; the upcoming shifts were not showing at first but now it appears on my dashboard.
<img width="1436" alt="Screenshot 2025-02-08 at 10 16 54 PM" src="https://github.com/user-attachments/assets/d96bd072-3505-46ab-9988-4614fae9c369" />
<img width="1437" alt="Screenshot 2025-02-08 at 10 16 22 PM" src="https://github.com/user-attachments/assets/e7454fe2-5d73-4e33-88e6-68c272784f7b" />
<img width="1440" alt="Screenshot 2025-02-08 at 10 29 21 PM" src="https://github.com/user-attachments/assets/57358d84-017b-44ce-88b2-a6679a4e243a" />

kate-holdener commented on client_app/src/components/volunteer/Impact.js:
```diff
@@ -0,0 +1,66 @@
+    const endpoint = `${SERVER}/shifts?filter_end_before=${time_now}`;
+    const header = getAuthHeader();
+
+    console.log("Fetching impact data from:", endpoint); // Debugging
```
```suggestion
```
I suggest removing debugging statements.

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -3,19 +3,26 @@ import { SERVER } from "../../config";
-      .catch((error) => console.log(error));
-  }, []);
+      .then((response) => {
+        console.log("Response Status:", response.status); // Debugging
```
```suggestion
```
I suggest removing debugging statements

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -3,19 +3,26 @@ import { SERVER } from "../../config";
+        return response.json();
+      })
+      .then((response) => {
+        console.log("Fetched Shifts Data:", response); // Debugging
```
```suggestion
```

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -3,19 +3,26 @@ import { SERVER } from "../../config";
+        console.log("Fetched Shifts Data:", response); // Debugging
+        setData(response);
+      })
+      .catch((error) => console.log("Error fetching shifts:", error));
```
```suggestion
      .catch((error) => console.error("Error fetching shifts:", error));
```

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -38,24 +45,31 @@ function Shifts(request_endpoint) {
   const time_now = new Date().getTime();
-  const shelters_endpoint = SERVER + "/shifts?filter_start_after=" + time_now;
+  const shelters_endpoint = `${SERVER}/shifts?filter_start_after=${time_now}`;
+  console.log("Upcoming Shifts Endpoint:", shelters_endpoint); // Debugging
```
```suggestion
```

kate-holdener commented on client_app/src/components/volunteer/Shifts.js:
```diff
@@ -38,24 +45,31 @@ function Shifts(request_endpoint) {
   const time_now = new Date().getTime();
-  const shelters_endpoint = SERVER + "/shifts?filter_end_before=" + time_now;
+  const shelters_endpoint = `${SERVER}/shifts?filter_end_before=${time_now}`;
+  console.log("Past Shifts Endpoint:", shelters_endpoint); // Debugging
```
```suggestion
```

kate-holdener reviewed (changes requested):
@Taktar Nice job! There are a few debugging statements you left in the code. I marked them inline with my suggested change. Please go through them and either accept the changes or respond (if you think they should stay).

Taktar commented:
<img width="1432" alt="Screenshot 2025-02-09 at 11 37 18 AM" src="https://github.com/user-attachments/assets/46578cee-e0c9-4a61-8c1e-5c51f9e4f9f8" />
checking the next morning to see if my shift from last night shows in the previous shifts section; it does! YAY

Taktar commented:
<img width="1437" alt="Screenshot 2025-02-09 at 11 52 16 AM" src="https://github.com/user-attachments/assets/e01dab75-4502-45ab-8ac7-30b5abecc238" />

kate-holdener reviewed (approved):
Great job!
